[Maurice Mok Bing Wei] iP

Maurice2n97 opened the pull request.

Maurice2n97 committed:
Implemented greet, echo, exit

Maurice2n97 committed:
Added remember commands, or list them out

Maurice2n97 committed:
added the Task class

Maurice2n97 committed:
Added functionality to Chatbot to Mark Done Tasks

Maurice2n97 committed:
Added Deadline and ToDo tasks.

Maurice2n97 committed:
Added Event.java and modified toString for Deadline.java

Maurice2n97 committed:
Implemented tracking of three types of task

Maurice2n97 committed:
got rid of some duplication

Maurice2n97 committed:
Added tracking for three tasks req for level 4.

Maurice2n97 committed:
Minor changes, set up automated testing

Maurice2n97 committed:
Create class DukeException

Maurice2n97 committed:
Added changes to enum

Maurice2n97 committed:
updated additional checks for lowerCase, UpperCase, handle empty command( e.g, todo)

Maurice2n97 committed:
no message

Maurice2n97 committed:
Trying to fix problem with UI testing

Maurice2n97 committed:
fixed UI-testing

Maurice2n97 committed:
Added functionality to delete Tasks, Updated test_cases

Maurice2n97 committed:
Added comments to code.

Maurice2n97 committed:
Extract out letterCode for all subclasses of Task.
Task Class and its subclasses do not have a coded letter for quickly identifying what type of Task it is.
Task Class now has a String variable inside which stands for the letter code, i.e 'T', for task  'D' for deadline
Enhances readability and people would know what the variable stands for, may allow for easier generalisation between Task classes.

Maurice2n97 committed:
Add feature to store the taskList

When the program closes, memory of the original taskList is lost.

Store in a text file, Use a Storage Class to fetch the data. Furthermore a Parser class will  parse the saved format.

This adds more functionality to our Duke Chatbot.

Maurice2n97 committed:
Extract out classes for UI and TaskList

Main method requires cleanup as it is being cluttered with too many methods.

Extracting out the useful methods for UI into the UI class and seperately implementing a Tasklist class would be more organised and enhance readability. Also it would conform to SLAP principles.

Maurice2n97 committed:
Extract date from Event description

For Deadline and Event classes, we need a way to parse and store the dates associated with these two classes.

Static methods in Parser allows for parsing based on the date. So far the only recognisable format is  yyyy-MM-DD. F
uture iterations could have more date formats.

This would facilitate future features which can categorise events based on their dates.

Maurice2n97 committed:
fix missing newline output

Upon writing to the file there was no newline seperator between entries to the duke.txt

Having newline makes it easier to read the txt file and the bug would also cause problems reading from duke.txt

Maurice2n97 committed:
Handle file or directory is missing

There might be a problem when the program is run on other computers where they do not have the directories or files needed by duke. In this case we need to handle it by creating the files or directories for them.

Additionally added Duke constructor where the missing file or directory case will be handled.

Maurice2n97 committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Parser.java

Maurice2n97 committed:
Fix bug of not converting between formats

There are two format for the string representation of Task, the readable format and the saved format. toString will return the more readable date format while the saved format is kept different

Maurice2n97 committed:
Shorten Code

enhance readability by using DateFormatter API in java to easily extract date out from String instead of regex

Maurice2n97 committed:
Flesh out classes before refactoring

Refactoring code before creating new Command class

Maurice2n97 committed:
Organise into packages and implement TaskList class wrapper

Organise the classes into packages as there are too many classes and some needs to be logically grouped together. The new TaskList class also acts as a wrapper over TaskList and seperates it out from the main.

Organising into packages is neater and hence we can find files fast. It also helps with maintainability. The TaskList class adheres more strongly to OOP design.

Maurice2n97 committed:
Fix bug IndexError when deleting entry from TaskList

running text-ui test shows a bug when deleting the last indexed entry from TaskList. A previous change to Tasklist's delete method did not subtract 1 from the indexToDelete when calling arraylist.remove.

Maurice2n97 committed:
Minor changes to text-ui-test file

Minor changes, add some line breaks for readability

Maurice2n97 committed:
Create Command class and its subclasses

Representing the possible Actions that Duke Chatbot can take after parsing the command.

Breaking into Parser Class would allow better readability and maintainability by breaking the program into smaller modules.

**TODO ** Parser class would take a string and depending on the keyword parsed, will extract out the string and create the appropriate task/ command to pass to Main.

Maurice2n97 committed:
Rename packages to smallcase.

this is to suite java package naming conventions

fix some small errors as well.

Maurice2n97 committed:
Refactor Parser to parse input

made Parser handle parsing input into the correct Command. This is so the logic flow is clearer and all the bulk of the string processing is located in the Parser class.

Maurice2n97 committed:
Finish up parsing for exit and List Command.

Checkpoint in case I need to revert back the changes.

Maurice2n97 committed:
Seperate out Parser Logic from Main.

Adhering to true OOP principles, restructure the code so that it is better organised . The Parser class handles most of the logic of reading input String and figuring out what Command to return.
Command.execute() carries out the calling of the UI and the storage. There is more work to be done to seperate out the Ui Logic from the Command class.

** TODO ** Seperate out the printing to the ui class. Also could seperate out Date Parsing logic from the Command Parser class . For future iterations/

Maurice2n97 committed:
Remove Comments and fix minor bug.

sc.hasNextLine() waits for the user, thus creating an issue where at the end of execution after inputting bye, the program still waits for the user input before exiting entirely.

fix : Deleted the need for a sc.hasNextLine() check.

Maurice2n97 committed:
Deleted redundant function

hasUserInput is not used.

Maurice2n97 committed:
Revert "Deleted redundant function"

This reverts commit 06708dbba8bfdb1727ae951a8e06f198b6a3ce10.

Maurice2n97 committed:
Delete redundant method in Ui

ui.hasUserInput is not used and hence deleted.

Maurice2n97 committed:
Add J-Unit Tests for Deadline.

Add JUnit tests to test Deadline. Check two methods for correctness, the toString method,
and the method for generating the saved String format used for storing the task in the file.

There could be instances where there is a typo that results in an incorrect String format.
This might lead to bugs when the tasks are being stored in another .txt file and fetched the next
time the app is open.

J-Unit tests help to automate the testing.

Let's add unit test to check for toString method correctness.
a unit test to check the getSavedString format correctness
and a unit test to check if Deadline throws Exception when
no date is able to be parsed from its by String.

Maurice2n97 committed:
Add  JavaDoc Comments and extract duplicate code

There is duplicate code in the toString method of various subclasses of Task. The common code can be extracted out under the Task class,
hence shortening the code. JavaDoc comments are needed to document what each class does and to make it easier to read.

Let's put JavaDoc comments for at least half of the classes, with the rest to be filled in at later dates.
Let's extract out the this.getLetterCode() function call  to the Task class method.

Maurice2n97 committed:
Fixed Coding Style

fixed coding style to adhere to the Coding Style standard and make the code appear to be written by one user.

Maurice2n97 committed:
Save checkpoint

Maurice2n97 committed:
Update Duke

Previously, Duke does not do case insensitive matching when finding tasks via keyword. Also, Duke throws an Exception when an empty line is entered to the chat bot

The case insensitive matching would be better ideal for searching .Second point  would not be good as the user may accidentally hit enter and cause the program to terminate.

Let's make Duke ignore empty return lines
Let's make Duke convert all search input to lower case, and target string to search will be lower case as well.

Maurice2n97 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fixed Coding Style

Maurice2n97 committed:
Merge branch 'branch-A-javadoc'

* branch-A-javadoc:
  Add  JavaDoc Comments and extract duplicate code

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Ui.java

Maurice2n97 committed:
Extract out field isExit from Command class

Previously, the command class execute method would, after execution, return a boolean indicating
if the program should terminate or not. This is harder to read as one would have to look
for the meaning of the returned boolean value from execute.

 iI the boolean variable were to be separated out as a  field within the command c it would be easier.

Let's modify execute to be a void method, while the Command
class contains the attribute isExit  and method shouldExitt(),
which should inform the user if the command is telling the program to exit.

VRSoorya commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+    private String by;
+    private LocalDate dateOfDeadline;
+
+    public Deadline(String description,String by) {
```
All classes and public methods need header comments, please check here: https://se-education.org/guides/conventions/java/intermediate.html#comments

You might have missed the space after "," in the input parameters

VRSoorya commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * main class containing the Duke Chatbot main logic.
+ */
+
```
Should there be an empty line here after the comment?
```suggestion

```

VRSoorya commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,327 @@
+     * @param inputCommand String to be parsed.
+     * @return an exit Command.
+     * @throws DukeException the string is not of the correct Exit Command format.
+     */
```
This is a good JavaDoc example!

VRSoorya commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,327 @@
+
+    /**
+     * Parses the input to see if it is of the DeleteCommand format,
+     * if so returns the Command to mark task as done.
```
You mean returns the command to delete?

VRSoorya commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    /**
+     * set the state of the task to be done.
```
"Sets" instead of set?

VRSoorya commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,327 @@
+
+    public Command parseCommand() throws DukeException{
+        String command = getKeyWord(inputCommand).toLowerCase();
+        Task t;
```
Is this line not used?

VRSoorya commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+     * @return length of the list
+     */
+
+    public int size() {
```
Please use a verb for method names like "getSize"

VRSoorya commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,29 @@
+        Task task = tasks.delete(indexToDelete);
+        try {
+            storage.saveTasks(tasks);
+        } catch ( IOException err) {
```
Please remove space after the bracket, should be: ...(IOException err) ...

VRSoorya commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,36 @@
+            this.dateOfEvent = Parser.parseDate(dateString);
+            this.at = at;
+        } else {
+            throw new IllegalArgumentException("Sorry Unable to Parse Date for Event. Did you put in yyyy-mm-dd format?");
```
Please use 2 lines instead for long statements like you have done in the last test in DeadLineTest.java

VRSoorya reviewed (commented):
Hey Maurice, overall your code is very clean and easy to follow, great job! I like how you used classes for each command and handled your exceptions too. 

JavaDoc comments need reviewing and can be more consistent throughout your codebase. Some minor edits suggested to adhere to coding standards as well. 

JavaDoc expectations can be found here: https://se-education.org/guides/conventions/java/intermediate.html#comments

VRSoorya commented:
I also noticed your Parser.java file is rather long, is there a way to refactor it? Perhaps one way would be to check the validity of the inputs in the Command classes?

Maurice2n97 committed:
Integrate Gradle for automating builds

Automated testing would ease the workflow of deploying and testing code.

Gradle support is provided for checkstye and JUnit testing

**TODO** add support for buidling JAR files using ShadowJar in Gradle

Maurice2n97 committed:
fix Checkstyle errors

Code is messy with many violation of the coding standards

make code neater and in line with JLS coding standards, using gradle checkstyle

Maurice2n97 committed:
Create simple GUI for chatbot.

basic GUI chatbot window to type in commands. Not a finished article yet. Save Checkpoint here before tinkering.

Maurice2n97 committed:
Add more functionality to GUI

Add pictures, dialog boxes, spacing.

Maurice2n97 committed:
Separate out Ui logic form Ui design

The application class is getting too bulky

Lets Use controller view design to separate out business logic.

Maurice2n97 committed:
Connect GUI to main logic

GUI now is able to output responses from Duke Chatbot.

Maurice2n97 committed:
Set Insets for components in GUI

Alter FXML files to set margins around the images and text labels within dialog boxes.

Looks nicer overall.

Maurice2n97 committed:
Set GUI to be resizeable

Resizeable GUI that grows the comment bar and the scroll pane to fit the window.

Maurice2n97 committed:
Got rid of Duke Welcome Message

When GUI displays the welcome message, there is some formatting issues with regards to how the whitespaces are sized relative to the "_" characters. Decide to remove it for cleaner UI.

Maurice2n97 committed:
Refactor Code for Parser

Ensures code is within code quality standards. Split parseCommand into two seperate steps.
First step checks for match, second step extracts out the match. Simplify code for find Command and
todo command as well.

Maurice2n97 committed:
Refactor Classes

Refactor classes to satisfy code quality standards

Maurice2n97 committed:
Extract out error message

Extract the error message to a private final static field within deadline and event classes.

Makes it easier to locate within the code, makes it neater as well.

Maurice2n97 committed:
Add Assertions for Parser,TaskList

Provide a sanity check that the index to access in the task list is not negative.
Provide a sanity check that the inputCommand should not be negative.

Maurice2n97 committed:
Merge pull request #2 from Maurice2n97/A-CodeQuality

A-CodeQuality

Maurice2n97 committed:
Merge branch 'master' into A-Assertions

* master:
  Extract out error message
  Refactor Classes
  Refactor Code for Parser

# Conflicts:
#	src/main/java/duke/Parser.java

Maurice2n97 committed:
Merge pull request #3 from Maurice2n97/A-Assertions

Add Assertions for Parser,TaskList

Maurice2n97 committed:
Create new  CommandResult class

Stores program execution state inside CommandResult , which is returned from executing a command. This is in contrast to previously altering a field in the ui to store the messages, which is dangerous as changing the state causes some coupling of the code.

Maurice2n97 committed:
Refactor storage class

Shorten functions by breaking them up into smaller functions.

Enhances readability and makes the happy path more prominent. Apply more SLAP.

Maurice2n97 committed:
Create new class Storage Parser

Extract method refactoring done on Parser to extract out methods used for parsing the
encoded task string from database.

This is so that it is neater and Parser class will not be responsible for doing different unrelated things.
Parser will seperate out to CommandParser and StorageParser. Command Parser will just focus on
parsing Commands.

Maurice2n97 committed:
Handle ArrayListOutOfBounds Exception

Handle the case that the index is out of bounds : so as to prompt the user again for another index.

Ensures the program doesnt quit at runtime if user mistakenly enters a wrong index

Maurice2n97 committed:
Organize into packages

Organize classes into packages makes it easier to search for them and reduces clutter.

Maurice2n97 committed:
Minor changes to improve code quality

Put call to static field storage as CLASS_NAME.storage rather than INSTANCE.storage

Maurice2n97 committed:
Add feature sorting of tasks

Add feature which automatically sorts the tasks via the dates. ToDo tasks
 have maximum dates, so they will be pushed to the end of the list.
The list automatically sorts itself whenever it is changed ( or when the model TaskList is initialized).

Sorting helps the user track dates easily instead of the dates being jumbled around.

Maurice2n97 committed:
Merge pull request #4 from Maurice2n97/C-Sort

Add feature sorting of tasks

Maurice2n97 committed:
Set up CI/CD

 add gradle.yml file for GitHub to run continuous integration

Maurice2n97 committed:
Update gradle.yml file

Maurice2n97 committed:
Break down method add

Extracted out add and sort functions for taskList into separate functions. Used insertIntoSorted position to combine both functions together.

Applied SLAP. now the add is no longer doing two things at the same time.

Maurice2n97 committed:
Fix bug regarding input with whitespace

Input with leading whitespaces are not trimmed.
Also renamed packages to lower case to match java standards.

Maurice2n97 committed:
fix bug regarding invalid Date

There might be invalid Dates for example 2021-20-20 is not a valid date in the calender.
Causing an unhandled DateTimeParseException to not be handled during runtime.
Created a new exception DukeDateTimeException to catch errors from Date Parser.

What could be done in the future. A dependency injection framework for initializing Time - based tasks like deadline.

Maurice2n97 committed:
Add some JUnit Tests

Increase code coverage for Parser classes.

Maurice2n97 committed:
Handle Error when storage file cant be created

The previous code throws a runtime error when there is an IO Exception in the case where
the txt file to store the tasks does not exist in the directory.

This could be not informative to the user who needs to diagnose the problem.

Lets-
* Handle the IO exception within the controller class main window
* Print out a message telling the user to close the app.
* Ask user to check the directory in qn and see if the file cant be created insdie

Maurice2n97 committed:
Add Colors to components via CSS

The default colours look dreary. Photo may have royalties associated with it.

Replacing them with custom colours would make it look nicer. Replace the photo with non royalty photo, with credit to (photographer)

Lets -
* add java FX CSS skins to button and viewport
*Change pictures used before release to no-royalty stock images

Maurice2n97 committed:
Setup README, Add product image

It is time to publish the second version

Lets -
*include the User guide for people to use
* Update the profile pic for showcase

Maurice2n97 committed:
Set theme jekyll-theme-cayman

Maurice2n97 committed:
Fix typos in README

small typos are found

hinders readability of user guide

Lets -
*Update the user guide with a more correct version

Maurice2n97 committed:
Merge branch 'master' of https://github.com/Maurice2n97/ip

* 'master' of https://github.com/Maurice2n97/ip:
  Set theme jekyll-theme-cayman

Maurice2n97 committed:
Refactor to match coding standard

There are minor coding standard violations

This makes it hard to standardise it to make it seem like it is written by only one person

Lets
- Make extra javadocs
- Ensure code follows the java coding standards

Maurice2n97 committed:
Update README to get rid of errors

Some inaccuracies have to be corrected.

Maurice2n97 committed:
Update README to get rid of more errors.

Maurice2n97 committed:
Update README

To fix spelling errors
